modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rel/IgniteMergeJoin.java

```diff
@@ -173,24 +184,34 @@ else if (isPrefix(rightCollation.getKeys(), joinInfo.rightKeys))// preserve righ
         Map<Integer, Integer> rightToLeft = joinInfo.pairs().stream()
             .collect(Collectors.toMap(p -> p.target, p -> p.source));
 
-        List<Integer> collationLeftPrj = new ArrayList<>();
+        List<Integer> collationLeftPrj = new ArrayList<>(collation.getKeys().size());
 
         for (Integer c : collation.getKeys()) {
             collationLeftPrj.add(
                 c >= rightOff ? rightToLeft.get(c - rightOff) : c
```

Comment:
looks like there is another bug: we can't preserve collation if required collation contains right table's field outside of join pairs

Reply:
i extend tests but have no reproducer for this assumption ...